### PR TITLE
Generic state synchronization: runtime-schema state sync + phase demo (PAI-1096)

### PR DESCRIPTION
**Pull request by dokterbob**
## Summary

Adds a **generic, runtime-JSON-Schema-driven state synchronization primitive**: any field in the LangGraph state schema can be wired to a typed, reactive frontend widget that reads live thread state and writes back through the graph — demonstrated end-to-end with a `phase` field (`research`/`draft`/`review`) shown as a dropdown above the chat. Motivated by PAI-1096 (purposepilot phase sync), built generic because the runtime-schema machinery is also the foundation for future dynamic UI (tool-arg forms, interrupt payloads).

### How it works

- **Read path:** `createStateSync({ stream, client, assistantId })` fetches the assistant's live `state_schema` (`assistants.getSchemas`), parses it into typed field descriptors (`schema.ts` — handles `$defs`/`$ref`, nullable `anyOf`, never throws), and vends reactive `field(name)` bindings backed by `stream.values`. `StateField.svelte` renders schema-appropriate widgets (enum→select, boolean→toggle, string/number→input, snippet escape hatch). If the schema is null/unavailable, the widget hides and chat works normally.
- **Write path (graph-mediated only):** `field.set(v)` submits a state-only run (`stream.submit({ phase: v }, { optimisticValues })`). The backend's `phase_gate` middleware applies the state delta and jumps to `end` without invoking the LLM — instant, validated, and fully visible to `useStream`. `threads.updateState` is documented as an expert escape hatch only (no validation, 409s during runs, invisible to the hook).
- **AI writes:** a `change_phase` tool updates state via `Command`; the current phase is exposed to the model as a dedicated system message.

### Backend shape

The graph is a flat `langchain.agents.create_agent` with middleware (`2921cc2`), not an outer `StateGraph` wrapping a react-agent subgraph:

- **`phase_gate`** is a `@before_agent` hook. It defaults a missing `phase` to `DEFAULT_PHASE`, and **rejects an invalid one by raising** — deliberate fail-fast, since the bad value is already checkpointed by the time the hook runs, so message-only submits keep failing until a valid write lands. That surfaces a caller bug loudly instead of silently coercing it away, and the thread stays recoverable (`test_invalid_phase_rejected_then_recoverable`).
- It also ends the run without a model call for state-only submits (marked via `configurable.state_only_submit`, passed per-invocation rather than through graph state so it is never checkpointed) and when the last message isn't a `HumanMessage`.
- **`PromptMiddleware`** builds the model input (system prompts + injected phase message + greeting + state messages).

### Concurrent tool writes: `phase` needs a reducer

Two `change_phase` tool calls in **one assistant message** killed the run with `INVALID_CONCURRENT_GRAPH_UPDATE` — reported by @aryadhruv, fixed in `3d03785`:

```
At key 'phase': Can receive only one value per step.
Use an Annotated key to handle multiple values.
```

`ToolNode` runs tool calls concurrently via `executor.map` and merges only `Command`s targeting the **parent** graph, so sibling calls each write the `phase` channel in the same superstep. `phase` had no reducer; `messages` survived only because it inherits `add_messages`.

Fixed with a generic `last_value` reducer (new `svelte_langgraph/reducers.py`), applied as `phase: Annotated[Phase, last_value]`. It lives in its own module because **every** future tool-writable synced field has this same hazard. `executor.map` preserves `tool_calls` order, so "last" is the last tool call in the message.

Verified the reducer changes nothing else: the derived JSON schema is **byte-identical** with and without it — checked against the `_state_jsonschema` path the frontend actually consumes via `getSchemas()`, not just Pregel's own methods — so no frontend change was needed. An unset `phase` still leaves the channel empty (`EmptyChannelError`, as before), so `phase_gate`'s `DEFAULT_PHASE` path is untouched. Also confirmed it genuinely *folds* at 3- and 4-way writes with reversed and non-alphabetical permutations, not just the pairwise case.

### Notable fixes bundled (found while making E2E pass locally)

- **`slow_mock.py` emits OpenAI-spec streamed tool-call deltas:** stock mockai omits `index` and repeats `id`/`name` per delta, so langchain-openai accumulates empty tool args (tool fires with `{}`).
- **`slow_mock.py` fixes mockai's multi-tool-call validator:** `verify_structure` rejects *every* array-shaped `output`, because pydantic parses the array into a `FunctionOutputs` RootModel and the validator tests `isinstance(output, list)` — which a RootModel is not. The serving path already supports multi-call output; only the validator is wrong. Since `dependencies.py` validates the whole file at startup, one such entry stops the mock booting at all — which blocked testing parallel tool calls entirely. The patch self-checks at import so it fails with a self-describing error rather than an opaque pydantic one if upstream shifts, and `ai-mock` is now pinned `<0.4` since we depend on its internals. Upstream is unmaintained (last commit 2025-03-07; our issue #10 unanswered; a community PR fixing this exact defect unmerged ~1 year) — tracked as **SLG-129** to fork or replace it.
- **frontend `clean` now wipes `.svelte-kit`:** a stale `generated/server/internal.js` baked an old `__sveltekit_<version_hash>` into the SSR bundle while the client pass computed a fresh one → split build → total hydration failure locally (follow-up to #254).
- **Playwright webServer wait regexes tolerate ANSI colors** (moon allocates a PTY locally; oidc-provider-mock/vite colorize their ready lines) + `reuseExistingServer` outside CI.

### Docs

- `apps/frontend/README.md`: usage, write-path rationale, degraded mode, caveats (queued submits, checkpoint branching rewinding synced fields, **and now: concurrent tool writes need a reducer**), `updateState` hazards.
- `docs/spikes/state-update-sse.md`: cross-client live-sync research — `state_update` SSE fires only on explicit `updateState` (conditional GO on langgraph-api, NO-GO on Aegra).
- `docs/upstream/langchain-svelte-expose-history-mutate.md`: filed as langchain-ai/langgraphjs#2589.

## Test plan

- **Backend: 610 unit tests** — state-only submit (asserts zero LLM calls), tool `Command` flow, invalid-phase rejection + recovery, schema enum exposure, regenerate re-execution, and the new parallel-tool-call test, which pins `phase == "review"`, both `ToolMessage`s correlated to their originating `tool_call_id`, **and tool-call ordering** — so a future langgraph-prebuilt change that flips the winner fails loudly rather than silently.
- **Frontend: 220 unit tests** — schema parser edge cases, stateSync reactivity/degraded mode/dev warnings, and a reactive select test covering the browser option-dirty-flag trap.
- **E2E: 43 tests**, including `e2e/src/phase.spec.ts` (4): manual phase change is state-only (no AI message, survives reload), phase visible to the model, AI-driven `change_phase` updates the dropdown, and AI-driven **parallel** phase changes resolve last-write-wins with a reload to confirm the server committed it.
- CI green on `3d03785`: `CI Job`, `Docker Build`, CodeQL (python + js/ts), `qlty fmt`, `qlty coverage` (65.0%, +5.0%) and `qlty coverage diff` (92.1% vs 70% threshold) all pass. `qlty check` reports "Build errored" — an infrastructure-level error, not a code finding; a `qlty check` status also posted "No blocking issues".

### Known follow-ups

- Three references to the removed **entry router** design survive in `apps/frontend/README.md` and `stateSync.svelte.ts`'s doc comment; they should say `phase_gate` middleware.
- **Pre-existing optimistic-UI race, now reachable:** `stateSync.svelte.ts` treats the UI's own write as authoritative until round-trip, so changing the dropdown while the model emits two `change_phase` calls can settle on a different value than the UI showed. Previously that scenario crashed the run loudly; now it resolves quietly.
- **`e2e_responses.json` trigger strings must stay unique** — mockai's `_verify_responses` doesn't check plain-string `input` collisions and `find_matching_or_none` returns the first match, so a future entry could silently shadow an existing one.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/frontend/src/lib/langgraph/stateSync.svelte.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+					}
+				}
+				stream.submit(
+					{ [name]: v },
+					{
+						optimisticValues: (prev: Record<string, unknown>) => ({
+							...prev,
+							[name]: v
+						})
+					}
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/frontend/src/lib/langgraph/stateSync.svelte.ts:215]: Investigated: the SDK's `StreamOrchestrator.submit()` already hardcodes `multitaskStrategy: "enqueue"` server-side whenever the stream is busy, regardless of what's passed in `submitOptions` — matching this file's own existing doc comment ("enqueues the submission via useStream's built-in queue"). No code change needed; added a one-line comment at the `stream.submit()` call site citing this.

File: apps/frontend/src/lib/langgraph/schema.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
+/**
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/frontend/src/lib/langgraph/schema.ts:1]: Investigated: this is qlty's bundled prettier disagreeing with the project's own prettier config, not an actual formatting issue. Running `qlty fmt` collapses a multi-line union type that the project's real prettier (`moon frontend:format`) deliberately keeps multi-line — applying qlty's suggestion would break this repo's actual formatting standard. No change made.

File: apps/frontend/src/lib/components/StateField.svelte [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			<!-- Function binding drives the select's value property directly. Per-option
+			     `selected` attributes would break here: once the user has picked an option,
+			     the browser ignores attribute changes on it (dirty flag), so later
+			     programmatic state changes could desync the visible selection. -->
+			<select
+				id={inputId}
+				aria-label={accessibleName}
+				bind:value={() => (field.value as string) ?? '', (v) => field.set(v)}
+				class="border-border bg-card text-foreground focus:ring-ring rounded border px-2 py-0.5 text-xs focus:ring-1 focus:outline-none"
+			>
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/frontend/src/lib/components/StateField.svelte:60]: Investigated: `bind:value={() => ..., (v) => ...}` is Svelte 5's documented "function bindings" syntax (specially parsed by the compiler as a get/set pair, not a JS comma expression). All 13 tests in StateField.svelte.test.ts pass, including one that was empirically verified to fail if this exact binding is broken — confirming it works correctly at runtime. No change made.

File: e2e/src/phase.spec.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+		// Optimistic update: dropdown reflects the new value immediately.
+		await expect(chat.phaseSelect).toHaveValue(targetValue);
+
+		// Bounded wait (~3 s) — the state-only run must NOT produce any chat message.
+		await page.waitForTimeout(3000);
+		expect(await aiGroups.count()).toBe(countBefore);
+
+		// Input re-enabled confirms the run completed cleanly.
+		await expect(chat.textInput).toBeEnabled({ timeout: 15000 });
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply e2e/src/phase.spec.ts]: Fixed in "Remove flaky sleep and fix cross-spec thread races in phase.spec.ts" (now cb1ef03 after rebase): the fixed waitForTimeout(3000) is gone — the test now waits for the completion signal and asserts the AI message count is unchanged.

File: apps/frontend/src/lib/langgraph/schema.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+ * Accepts `unknown` so callers can pass the SDK's `GraphSchema.state_schema`
+ * directly without needing a cast (the SDK's `JSONSchema7` and our local one
+ * have different structural types for `properties` due to the JSON Schema spec
+ * allowing boolean sub-schemas, e.g. `{ "additionalProperties": false }`).
+ *
+ * Returns `{ status: 'unavailable' }` when:
+ *  - `schema` is `null`, `undefined`, or not an object
+ *  - The schema has no `properties` (not an object schema)
+ *  - An unexpected exception is thrown during parsing
+ *
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/frontend/src/lib/langgraph/schema.ts:191]: Fixed in "Reduce parseFieldSchema complexity, fix parseObjectSchema JSDoc" (now 4975daf after rebase): the JSDoc now explicitly documents that a schema with no properties returns { status: 'ok', fields: {} } and enumerates the actual 'unavailable' conditions.

File: e2e/src/phase.spec.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+ * (no message, no LLM call) via stream.submit.  The backend entry router applies
+ * the phase and exits.  The AI can also change phase via the `change_phase` tool;
+ * the update reaches stream.values when the agent node completes its superstep
+ * (root-level `values` events are per OUTER superstep — without streamSubgraphs
+ * the tool's update is not visible mid-generation).
+ *
+ * These tests share backend thread state for the same authenticated user, so they
+ * run serially to prevent races between concurrent phase mutations.
+ */
+test.describe.configure({ mode: 'serial' });
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply e2e/src/phase.spec.ts:23]: Addressed across two commits. phase.spec.ts itself stopped reading shared state via gotoFreshThread() (cb1ef03), and the remaining direction of the leak — other specs picking up phase-test threads through getOrCreateThread()'s latest-idle reuse — is closed in 0f62122: gotoFreshThread() is hoisted into fixtures/backend.ts and used by chat.spec.ts's content-sensitive Edit/Regenerate/Cancellation blocks. Note CI runs with workers: 1, so the failure mode was sequential thread inheritance rather than true concurrency; the fix covers both.

File: apps/frontend/eslint.config.js [hunk 7]
```diff
@@ -3,34 +3,37 @@ import js from '@eslint/js';
...
-	includeIgnoreFile(gitignorePath),
-	js.configs.recommended,
-	...ts.configs.recommended,
-	...svelte.configs.recommended,
-	prettier,
-	...svelte.configs.prettier,
-	{
-		languageOptions: {
-			globals: { ...globals.browser, ...globals.node }
+export default (async () => {
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/frontend/eslint.config.js]: Resolved: ESLint 9 flat config does support a Promise default export, so this wasn't breaking moon frontend:lint — but the config was reworked anyway and now exports a synchronously-resolved value via top-level await. It has since been further hardened in #260 (653538f): sibling lookups prefer the config file's own directory with a cwd fallback, and the svelte.config.js import degrades gracefully when build-only deps aren't installed (which is what was actually failing the qlty check gate).

File: apps/backend/src/svelte_langgraph/tools.py [hunk 8]
```diff
@@ -10,5 +24,21 @@ async def get_weather(city: str) -> str:
...
+    """Change the current working phase."""
+    validate_phase(phase)
+    return Command(
+        update={
+            "phase": phase,
+            "messages": [
+                ToolMessage(f"Phase changed to {phase}.", tool_call_id=tool_call_id)
+            ],
+        }
+    )
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/backend/src/svelte_langgraph/tools.py:35]: Investigated: confirmed langgraph 1.0.7's default handle_tool_errors already converts an out-of-enum phase argument into a graceful error ToolMessage before change_phase's body runs (pydantic rejects it against the Literal schema first and it's wrapped as a ToolInvocationError), so the specific 'invalid tool argument' scenario doesn't crash the run. However, a bare ValueError from validate_phase inside the tool body — reachable if the Literal and VALID_PHASES ever drift apart — is re-raised by the default handler and does fail the whole run (empirically reproduced by monkeypatching validate_phase). Fixed defensively in 849ccaf: change_phase catches the ValueError and returns a Command with only an error ToolMessage (no state update), plus regression tests for both paths (test_change_phase_tool_value_error_becomes_error_tool_message, test_change_phase_tool_invalid_arg_from_llm_is_graceful).

File: e2e/src/chat.spec.ts [hunk 9]
```diff
@@ -107,6 +107,89 @@ test.describe('Edit message', () => {
...
+			.filter({ has: page.locator('.prose') })
+			.filter({ hasText: question })
+			.last();
+		await expect(aiMessage).toBeVisible();
+		// Wait for streaming to finish — the regenerate button is unavailable while streaming.
+		await expect(chat.textInput).toBeEnabled();
+
+		const threadId = new URL(page.url()).pathname.split('/').at(-1)!;
+		const headers = { Authorization: `Bearer ${capturedToken!}` };
+
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply e2e/src/chat.spec.ts:220]: Fixed in 0f62122: added expect(capturedToken, 'no authenticated request observed via page.route').toBeTruthy() immediately before the header construction, matching the guard pattern already used by the Cancellation test in this file.

File: apps/backend/src/svelte_langgraph/graph.py [hunk 10]
```diff
@@ -31,29 +42,73 @@ def get_checkpointer() -> Checkpointer:
...
+        logger.warning("Coercing invalid phase %r to 'research'", phase)
+        phase = "research"
+    return {"phase": phase}
+
+
+def _route_after_entry(state: AgentExtendedState) -> str:
+    """Route to agent if last message is from user; otherwise END (state-only submit)."""
+    messages = state.get("messages", [])
+    if messages and isinstance(messages[-1], HumanMessage):
+        return "agent"
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/backend/src/svelte_langgraph/graph.py]: Confirmed and fixed in 849ccaf. Reproduced exactly as described: after a failed/cancelled generation the checkpoint commits with a dangling HumanMessage (the entry node's superstep persists independently of the agent node), and a subsequent phase-only submit was routed back into the agent and re-invoked the LLM. State alone can't disambiguate this — the regenerate flow deliberately resubmits against a pre-agent checkpoint that looks identical. Fix: field.set() now sends config: { configurable: { state_only_submit: true } } and _route_after_entry checks it before the message-based fallback. The marker rides per-invocation RunnableConfig, never checkpointed state, so it can't leak into later chat submits. Regression tests: test_phase_only_submit_after_failed_generation_skips_llm (reproduces the bug pre-fix) and test_phase_only_submit_without_marker_still_regenerates (locks in regenerate behavior), plus a frontend test asserting the marker is sent.

File: e2e/src/phase.spec.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+	const session = await sessionRes.json();
+	const accessToken = session?.accessToken as string | undefined;
+	expect(accessToken, 'expected an accessToken on the Auth.js session').toBeTruthy();
+
+	const threadRes = await page.request.post(`${LANGGRAPH_CONFIG.apiUrl}/threads`, {
+		headers: {
+			Authorization: `Bearer ${accessToken}`,
+			'Content-Type': 'application/json'
+		},
+		data: { metadata: {} }
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply e2e/src/phase.spec.ts]: Confirmed and fixed in 0f62122. Traced the exact failure path: every spec authenticates as the same OIDC test-user, the backend scopes thread search to that identity, and the ai-mock matches purely on message-offset content — so a thread left idle with phase='draft' would answer any later message with 'PHASE_VISIBLE: draft' instead of the echo that chat.spec.ts's Edit/Regenerate/Cancellation assertions depend on. gotoFreshThread() is now shared in fixtures/backend.ts (with the rationale documented) and used by those three describe blocks; specs that deliberately exercise the getOrCreateThread reuse/redirect flow are intentionally left on the plain /chat route since they don't assert on message content.

File: apps/frontend/src/lib/langgraph/schema.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/frontend/src/lib/langgraph/schema.ts:1]: Follow-up: root-caused properly and fixed in #260 rather than left as-is. qlty resolved its prettier through a caret range (this pnpm repo has no package-lock.json/yarn.lock for qlty to extract an exact version from), installing prettier 3.9.5 vs the repo's locked 3.6.2, whose union-type wrapping differs. .qlty/qlty.toml now pins version = "3.6.2" with extra_packages for the svelte/tailwind plugins at lockfile versions — the qlty fmt gate is green on #260 and this branch is rebased on top.

File: apps/frontend/src/lib/components/StateField.svelte [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+				class="border-border bg-card text-foreground focus:ring-ring rounded border px-2 py-0.5 text-xs focus:ring-1 focus:outline-none"
+			/>
+		{:else if field.schema.kind === 'number'}
+			<label class="text-muted-foreground text-xs" for={inputId}>{accessibleName}</label>
+			<input
+				type="number"
+				id={inputId}
+				aria-label={accessibleName}
+				value={field.value as number | undefined}
+				onchange={handleNumberChange}
```
[the remark this answers is not in the document]
Comment (dokterbob) [reply apps/frontend/src/lib/components/StateField.svelte:97]: Investigated: this doesn't reproduce in Svelte 5. A plain value={...} attribute on an input compiles to the set_value runtime helper, which explicitly coerces nullish — its last line is element.value = value ?? '' (svelte/src/internal/client/dom/elements/attributes.js) — so undefined never reaches the DOM as the string "undefined". Number inputs additionally sanitize any non-numeric string to '' per the HTML value-sanitization algorithm. Rather than adding a redundant coercion, I added regression tests in 40d716e (the number kind previously had none): a fresh thread with undefined value asserts the input renders empty, plus coverage for rendering a current value and committing a parsed number via field.set — all passing against the unchanged component.

File: apps/backend/src/svelte_langgraph/graph.py [hunk 13]
```diff
@@ -31,29 +42,88 @@ def get_checkpointer() -> Checkpointer:
...
     model = get_chat_model()
-    checkpointer = get_checkpointer()
 
-    agent = create_react_agent(
+    # checkpointer=False disables subgraph persistence entirely (None would inherit
+    # the parent's). With inherited checkpointing, resuming from a parent checkpoint
+    # (the frontend's regenerate flow) re-attaches to the subgraph's own completed
+    # namespaced checkpoints and replays the cached answer instead of re-running
+    # the model. Root state is checkpointed by the outer graph per superstep.
+    inner_agent = create_react_agent(
```
Comment (aryadhruv) [inline apps/backend/src/svelte_langgraph/graph.py]: Comment is quite obvious for langgraph, ain't ? Also this is just a sample graph - i don't think we need some explaining here why this piece of code is here.

**Review by aryadhruv — CHANGES_REQUESTED** [review]
Hmm, streaming is not working anymore 🤔

File: apps/backend/src/svelte_langgraph/graph.py (hunk 13 above)
Comment (dokterbob) [reply apps/backend/src/svelte_langgraph/graph.py]: Fair point — trimmed to a one-line breadcrumb in 39c6280. Kept the pointer to test_regenerate_reexecutes_model because checkpointer=False vs None is load-bearing here (with inherited checkpointing, regenerate replays the subgraph's cached answer instead of re-running the model), and that's easy to "clean up" without noticing.

**Comment by dokterbob** [comment]
Heads-up on the streaming regression your colleague noticed while testing: **confirmed and fixed** in 39c6280.

**Root cause**: this PR wraps the react agent (previously the root graph) as a subgraph node inside the outer entry-router graph, and LangGraph only emits events from inside subgraphs when the client requests them. The UI therefore received no LLM token events — answers popped in whole at run completion. Confirmed at the library level: bare agent streams 5 message events, wrapped agent with default options streams 1, wrapped with `subgraphs=True` streams 5 again.

**Fix**: all four agent-run submit sites (send, retry, edit, regenerate) now pass `streamSubgraphs: true`. State-only field submits don't need it (they route to END with no LLM call).

**Why CI didn't catch it**: the existing "partial content" e2e assertions matched the user's own message bubble (the mock echoes the input back, and user bubbles contain the same text). Added a `Streaming` e2e test that scopes to AI bubbles and asserts the echo's head renders while its tail doesn't yet exist and the input is still disabled — verified to fail without the fix and pass with it. Full e2e suite: 36/36 green with CI parity (workers=1).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: apps/backend/src/svelte_langgraph/graph.py [hunk 14]
```diff
@@ -31,29 +42,85 @@ def get_checkpointer() -> Checkpointer:
...
+    is checkpointed before this node runs, so raising here would leave the bad
+    value committed to thread state and every subsequent run would fail on it.
+    Strict validation lives in the change_phase tool, where a bad LLM argument
+    becomes a recoverable ToolMessage error instead of a failed run.
+    """
+    phase = state.get("phase") or "research"
+    if phase not in VALID_PHASES:
+        logger.warning("Coercing invalid phase %r to 'research'", phase)
+        phase = "research"
+    return {"phase": phase}
```
Comment (dokterbob) [inline apps/backend/src/svelte_langgraph/graph.py]: I'm not sure validation should be here. Rather, couldn't we rely on `AgentExtendedState` (perhaps with pydantic) to ensure that the phase is a valid value and leave it to the calling logic to do the validation?

File: apps/backend/src/svelte_langgraph/graph.py [hunk 15]
```diff
@@ -31,29 +42,85 @@ def get_checkpointer() -> Checkpointer:
...
+    outer.add_node("entry", _entry_node)
+    outer.add_node("agent", inner_agent)
+    outer.add_edge(START, "entry")
+    outer.add_conditional_edges(
+        "entry", _route_after_entry, {"agent": "agent", END: END}
     )
+    outer.add_edge("agent", END)
 
-    return agent
+    return outer.compile(checkpointer=get_checkpointer())
```
Comment (dokterbob) [inline apps/backend/src/svelte_langgraph/graph.py]: Is it absolutely necessary to use a subgraph here, or could we reach the same functionality without it?

File: apps/backend/src/svelte_langgraph/graph.py [hunk 16]
```diff
@@ -1,17 +1,27 @@
...
 
 from .models import get_chat_model
-from .tools import get_tools
+from .tools import VALID_PHASES, get_tools
+
+logger = logging.getLogger(__name__)
+
+
+class AgentExtendedState(AgentState):
+    phase: Literal["research", "draft", "review"]
```
Comment (dokterbob) [inline apps/backend/src/svelte_langgraph/graph.py]: This seems redundant with `VALID_PHASES`.

File: apps/backend/src/svelte_langgraph/tools.py [hunk 17]
```diff
@@ -10,5 +24,39 @@ async def get_weather(city: str) -> str:
     return f"It's always sunny in {city}!"
 
 
+def change_phase(
+    phase: Literal["research", "draft", "review"],
+    tool_call_id: Annotated[str, InjectedToolCallId],
+) -> Command:
+    """Change the current working phase."""
+    try:
+        validate_phase(phase)
```
Comment (dokterbob) [inline apps/backend/src/svelte_langgraph/tools.py]: Once again, I would suggest we use pydantic for validation.

**Comment by dokterbob** [comment]
@aryadhruv I've just done extensive manual validation and I can confirm that:
1. The LLM can change the phase.
2. The LLM receives any changes in phase steered from the UI (although perhaps we should add a message for the LLM so it knows what's up and/or can do a generation).

I also verified correct continuation after logout/login.

**Review by aryadhruv — CHANGES_REQUESTED** [review]
A few bugs - see below comments

**Comment by aryadhruv** [comment]
Sync works well !
LLM calls tool - vola - phase changed.

But... if LLM tries to phase change twice in same run - we break the graph.
image

**Review by aryadhruv — APPROVED** [review]
Tested for the changes and regression - working ! All green
